### PR TITLE
mon/PGMap: fix summary display of >32bit pg states

### PR DESCRIPTION
**Pull request by liewegas**
Fixes: https://tracker.ceph.com/issues/44052

### CODE REVIEW

File: src/mon/PGMap.h [hunk 1]
```diff
@@ -49,7 +49,7 @@ class PGMapDigest {
   pool_stat_t pg_sum;
   osd_stat_t osd_sum;
   mempool::pgmap::map<std::string,osd_stat_t> osd_sum_by_class;
-  mempool::pgmap::unordered_map<uint64_t,int32_t> num_pg_by_state;
+  mempool::pgmap::unordered_map<uint64_t,int64_t> num_pg_by_state;
```
Comment (tchaikov) [inline src/mon/PGMap.h]: no, i think the original version is correct. please see `PGMap::stat_pg_add()`, where we have
```c++
num_pg_by_state[s.state]++;
```

so the key of the map is state, and the value is the number of pg with that state.

File: src/mon/PGMap.cc [hunk 2]
```diff
@@ -57,14 +59,21 @@ void PGMapDigest::encode(bufferlist& bl, uint64_t features) const
...
+      encode((uint64_t)p.first, bl);
+      encode((uint32_t)p.second, bl);
+    }
   } else {
     uint32_t n = num_pg_by_state.size();
     encode(n, bl);
     for (auto p : num_pg_by_state) {
       encode((uint32_t)p.first, bl);
-      encode(p.second, bl);
+      encode((uint32_t)p.second, bl);
```
Comment (tchaikov) [inline src/mon/PGMap.cc]: as put in my another comment, this is not correct.

but, before a25221e55172c326b48dea6f08c16d700921864e, 
```
mempool::pgmap::unordered_map<int32_t,int32_t> num_pg_by_state;
// v1
```

after a25221e55172c326b48dea6f08c16d700921864e:

```
mempool::pgmap::unordered_map<uint64_t,int32_t> num_pg_by_state;
// v2
    uint32_t n = num_pg_by_state.size();
     ::encode(n, bl);
     for (auto p : num_pg_by_state) {
       ::encode((uint32_t)p.first, bl); // <<< buggy! should have been `int32_t`
       ::encode(p.second, bl);
     }
```

so i think we should instead encode the key (state) as int32_t to be compatible with v1 decoder. like
```c++
if (v >= 2) {
  ::encode(num_pg_by_state, bl);
} else {
  uint32_t n = num_pg_by_state.size();
  encode(n, bl);
  for (auto [state, num] : num_pg_by_state) {
    encode((int32_t)state, bl);
    encode(state, bl);
  }
}
```

**Comment by liewegas** [comment]
Oh, duh.. right.  I can fix the uint32_t -> int32_t, but I don't think that's the bug (and I'm not sure what that would actually affect?).  The problem I see is a master cluster has PGs with the failed_repair bit set but the 'ceph -s' view on the mon does does not show that state.

**Comment by liewegas** [comment]
Oh, i see the bug.

**Pull request merged**